python/raydp/tests/test_data_owner_transfer.py

```diff
@@ -0,0 +1,174 @@
+          ("wong", 50, 15122211523)]
+
+  rdd = s.sparkContext.parallelize(data)
+  out = s.createDataFrame(rdd, ["Name", "Age", "Phone"])
+  return out
+  
+
+@pytest.mark.xfail(reason="data ownership transfer feature is not enabled")
+def test_fail_without_data_ownership_transfer():
+  """
+  Test shutting down Spark worker after data been put 
+  into Ray object store without data ownership transfer.
+  This test should be throw error of data inaccessible after
+  its owner (e.g. Spark JVM process) has terminated, which is expected.
+  """
+
+  from raydp.spark.dataset import spark_dataframe_to_ray_dataset
+  
+  num_executor = 4
+
+  ray.shutdown()
+  raydp.stop_spark()
+
+  ray.init()
+  spark = raydp.init_spark(
+    app_name = "example",
+    num_executors = num_executor,
+    executor_cores = 1,
+    executor_memory = "800M"
+    )
+
+  df_train = gen_test_data()
+  # df_train = df_train.sample(False, 0.001, 42)
+
+  resource_stats = ray.available_resources()
+  cpu_cnt = resource_stats['CPU']
+
+  # convert data from spark dataframe to ray dataset without data ownership transfer
+  df = spark_dataframe_to_ray_dataset(df_train, parallelism=4)
+
+  # display data
+  df.show(5)
+
+  # release resource by shutting down spark
+  raydp.stop_spark()
+  ray.internal.internal_api.global_gc() # ensure GC kicked in
+  time.sleep(3)
+
+  # confirm that resources has been recycled
+  resource_stats = ray.available_resources()
+  assert resource_stats['CPU'] == cpu_cnt + num_executor
+
+  # confirm that data get lost (error thrown)
+  df = df.to_dask()
+
+
+def test_data_ownership_transfer():
+  """
+  Test shutting down Spark worker after data been put 
+  into Ray object store with data ownership transfer.
+  This test should be able to execute till the end without crash as expected.
+  """
+
+  from raydp.spark.dataset import spark_dataframe_to_ray_dataset
+  import numpy as np
+  
+  num_executor = 4
+
+  ray.shutdown()
+  raydp.stop_spark()
+
+  ray.init()
+  spark = raydp.init_spark(
+    app_name = "example",
+    num_executors = num_executor,
+    executor_cores = 1,
+    executor_memory = "800M"
+    )
+
+  df_train = gen_test_data()
+
+  resource_stats = ray.available_resources()
+  cpu_cnt = resource_stats['CPU']
+
+  # convert data from spark dataframe to ray dataset,
+  # and transfer data ownership to dedicated Object Holder (Singleton)
+  df = spark_dataframe_to_ray_dataset(df_train, parallelism=4, _use_owner=True)
+
+  # display data
+  df.show(5)
+
+  # release resource by shutting down spark Java process
+  raydp.stop_spark(del_obj_holder=False)
+  ray.internal.internal_api.global_gc() # ensure GC kicked in
+  time.sleep(3)
+
+  # confirm that resources has been recycled
+  resource_stats = ray.available_resources()
+  assert resource_stats['CPU'] == cpu_cnt + num_executor
+
+  # confirm that data is still available from object store!
+  df.show(5)
+
+  # sanity check the dataset is as functional as normal
+  df = df.to_dask()
```

Comment:
Can we use other function to test sanity, e.g. to_pandas? Dask is not our dependency.

Reply:
done

Reply:
ds. to_pandas() will not work as tody in Ray 1.9.0, it is as expeced ,since the master head of ray-project/ray lacks support for binary input, it only supports pyarrow.Table as the underlying type of the block, as shown below (python/ray/data/impl/arrow_block.py Line 141)

    def add_block(self, block: "pyarrow.Table") -> None:
        assert isinstance(block, pyarrow.Table), block   # THIS IS A PROBLEM!
        self._tables.append(block)
        self._tables_nbytes += block.nbytes
        self._num_rows += block.num_rows

For above hassle, my simple quicky fix will be:

    def add_block(self, block: Union["pyarrow.Table", bytes]) -> None:
        if isinstance(block, bytes):
            reader = pyarrow.ipc.open_stream(block)
            tb = reader.read_all()
        else:
            tb = block
        self._tables.append(tb)
        self._tables_nbytes += tb.nbytes
        self._num_rows += tb.num_rows

To sum up, I don't think it is a raydp problem, as we don't want to serialize Java object just yet, so in my view this to_pandas() issue needs to be solved at ray project upstream. For comparison, to_pandas_ref() works perfectly.

Therefore, I'll just sanity check on dataset functionality such as mean() for our unit test :)